### PR TITLE
HPCC-28675 Add findDropZonePlane()

### PR DESCRIPTION
**Pull request by wangkx**
Map from a host/ip to a known dropzone plane name.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by jakesmith — COMMENTED** [review]
@wangkx - please revisit this once https://github.com/hpcc-systems/HPCC-Platform/pull/16742 is merged.

**Comment by jakesmith** [comment]
> @wangkx - please revisit this once #16742 is merged.

@wangkx - https://github.com/hpcc-systems/HPCC-Platform/pull/16742 has now been merged.

**Comment by wangkx** [comment]
I see the isHostInPlane() in https://github.com/hpcc-systems/HPCC-Platform/pull/16742. Not sure it can be used here. If it returns false, how do I know whether it is because the plane has no "@hostGroup"/"hosts" (I should check the given host with the "localhost") or the given host does not match with any host in the plane (I should not check the given host with the "localhost")?  Any other changes in https://github.com/hpcc-systems/HPCC-Platform/pull/16742 should be used here?

**Comment by jakesmith** [comment]
> @jakesmith I see the isHostInPlane() in #16742. Not sure it can be used here. If it returns false, how do I know whether it is because the plane has no "@hostGroup"/"hosts" (I should check the given host with the "localhost") or the given host does not match with any host in the plane (I should not check the given host with the "localhost")? Any other changes in #16742 should be used here?

I don't really get the "localhost" case.. nothing should be using these functions with a localhost.
It sounds like "localhost" is being used as a stand-in for cases where planes don't have hosts in other parts of the code, but they should not be doing any host matching in that case.

For cases when a specific host has been passed in, then it should not ever be localhost, and this function can call isHostInPlane on each plane to see if it matches.

**Review by jakesmith — COMMENTED** [review]
@wangkx - see reply.

**Comment by wangkx** [comment]
@jakesmith please review my latest commit. I may use the following code to find a plane based on the host.
```
        if (!findHostedPlane(host, true, hostedPlane) && strieq(host, "localhost"))
            findPlaneNoHost("lz", nullptr, hostedPlane);
```

**Comment by jakesmith** [comment]
> @jakesmith please review my latest commit. I may use the following code to find a plane based on the host.
> 
> ```
>         if (!findHostedPlane(host, true, hostedPlane) && strieq(host, "localhost"))
>             findPlaneNoHost("lz", nullptr, hostedPlane);
> ```

Discussed offline.

Going forward, anything that is making a request to interact with a dropzone, should supply the dropzone name, not just the hostname (and path).
Those changes have largely already been made in other PR's some merged as I understand,
however, to cater for legacy client requests, we need to find the dropzone based on the hostname *and* path.
NB: in a cloud system with multiple dropzones, it is quite likely they will be non-hosted and mounted, so hostname is irrelevant, the only distiguishing factor is the path prefix.

So I think:
a) if dropzone name is supplied - use it directly, no lookup of dropzone involved.

b) if not, you need to find a matching dropzone, based on the host and prefix. (there should only be 1 unique match)
A request with "localhost" should be treated as unhosted, mounted off-node storage (the norm. in cloud case). 

I would add a:
```
IPropertyTree *findPlane(const char *path, const char *host, bool ipMatch)
```

treat incoming requests for "localhost" as null, and search all non hosted planes for a match.

**Review by jakesmith — COMMENTED** [review]
@wangkx - see comment following offline discussion.

**Comment by wangkx** [comment]
@jakesmith I noticed that the getDropZonePlanesIterator() is used inside the code of the findPlane() you drafted. Since the code can only be applied for dropzone, I am wondering if the name should be changed to findDropZonePlane(). Or, I may add a getPlanesIterator() as I did in https://github.com/hpcc-systems/HPCC-Platform/pull/16764/commits/3e5a2f7f634701a5fa66496393b9780a355973df and call it the findPlane(). Please advise.

File: dali/base/dautils.cpp [hunk 1]
```diff
@@ -111,6 +111,28 @@ IPropertyTree * getDropZonePlane(const char * name)
     xpath.appendf("[@name='%s']", name);
     return getGlobalConfigSP()->getPropTree(xpath);
 }
+IPropertyTree * findDropZonePlane(const char * path, const char * host, bool ipMatch)
```
Comment (jakesmith) [inline dali/base/dautils.cpp:114]: trivial/not new: unusual not to have new line between implementation methods in a cpp.
Can you add and to ones that proceed this new method?

**Review by jakesmith — APPROVED** [review]
@wangkx  - looks good.
You may want to address trivial comment, if so, please change and squash.

**Comment by wangkx** [comment]
@ghalliday Please review and merge.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**